### PR TITLE
Less dynamic groovy is case aled

### PR DESCRIPTION
**Pull request by aledsage**

### CODE REVIEW

File: core/src/main/java/org/apache/brooklyn/location/ssh/SshMachineLocation.java [hunk 1]
```diff
@@ -126,6 +126,7 @@
 import com.google.common.reflect.TypeToken;
 
 import groovy.lang.Closure;
+import static org.apache.brooklyn.util.JavaGroovyEquivalents. groovyTruth;
```
Comment (bostko) [inline core/src/main/java/org/apache/brooklyn/location/ssh/SshMachineLocation.java]: There is some extra character here.

File: utils/common/src/main/java/org/apache/brooklyn/util/JavaGroovyEquivalents.java [hunk 2]
```diff
@@ -66,8 +66,9 @@ public static String elvis(String preferred, String fallback) {
     public static String elvisString(Object preferred, Object fallback) {
         return elvis(asString(preferred), asString(fallback));
     }
+    @SuppressWarnings("unchecked")
     public static <T> T elvis(T preferred, T fallback) {
-        return groovyTruth(preferred) ? preferred : fallback;
+        return (T) fix(groovyTruth(preferred) ? preferred : fallback);
```
Comment (bostko) [inline utils/common/src/main/java/org/apache/brooklyn/util/JavaGroovyEquivalents.java]: I think we should get rid of groovy code on places like elvis.
This is a potential vulnerability.

**Comment by bostko** [comment]
https://github.com/aledsage/brooklyn-server/pull/3

File: utils/common/src/main/java/org/apache/brooklyn/util/JavaGroovyEquivalents.java (hunk 2 above)
Comment (bostko) [inline utils/common/src/main/java/org/apache/brooklyn/util/JavaGroovyEquivalents.java:71]: Having `fix` is concerning I'd rather drop GString that may have been supplied.
Also having different behavior is concerning.
```java
    public static String elvis(String preferred, String fallback) {
        return groovyTruth(preferred) ? preferred : fallback;
    }
//..
// and a fixed version
//..
    public static <T> T elvis(T preferred, T fallback) {
        return (T) fix(groovyTruth(preferred) ? preferred : fallback);
    }
```

Comment (aledsage) [reply utils/common/src/main/java/org/apache/brooklyn/util/JavaGroovyEquivalents.java:71]: I agree the `fix()` is concerning (returning something different than what is passed in is not expected behaviour in groovy). However, if subsequent use of the value expects to cast it to `java.lang.String`, then we want to convert it from a `GString` to a `String` to let the subsequent java code not throw ClassCastExceptions.

Not quite sure I follow the comment about difference in behaviour. The `elvis(String, String)` will always return a string. The `elvis<T>(string,string)` will return a string, or if passed a `GString` will also return a string.

Comment (bostko) [reply utils/common/src/main/java/org/apache/brooklyn/util/JavaGroovyEquivalents.java:71]: @aledsage can we have a test with elvis(new GString(), new GString())?

The second, you are right,  I didn't check that it is `String` and do not need `fix`

Comment (aledsage) [reply utils/common/src/main/java/org/apache/brooklyn/util/JavaGroovyEquivalents.java:71]: @bostko done, and squashed into existing commits.

**Pull request merged**